### PR TITLE
[Upgrade Phase-2 Pixel 3D Digitizer] Fix misevaluation of charge when migrating due to diffusion

### PR DESCRIPTION
**Pull request by claralasa**
#### PR description:

Before this fix, more than half of the carriers were migrated due to diffusion. This made the residuals to change their sign in the region enabled for charge migration. In addition, a factor in the evaluation of migrated charge was missing.

Nevertheless, these fixes are barely noticeable due to the small allowed charge migration region (0.4 um) and so **no changes in the plots are expected**. In order to visualize the effect, such a region has been tested increasing up to 5 um. Just used for this study, **the value in the code stays as 0.4 um**. In the plot below, the charge per digi is shown. As can be observed, it does not change after the fix with the baseline of 0.4um.

FixedBugs_2

There are also some changes in the provisional validator used to cross-check the digitizer (which is not included within the validation framework).

This PR only concerns Upgrade (@duartej , @emiglior , @skinnari )

#### PR validation:

- Passed the runTheMatrix.py -l limited -i all --ibeos tests, don't know if they are relevant because I'm not sure if they were tested against 2026D65 geometry (or the current geometry with 3D pixels in barrel).

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No.

### CODE REVIEW

File: SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.cc [hunk 1]
```diff
@@ -158,56 +158,45 @@ std::vector<DigitizerUtility::EnergyDepositUnit> Pixel3DDigitizerAlgorithm::diff
...
   std::vector<float> newpos({pos_moving[0], pos_moving[1], pos_moving[2]});
   float distance_edge = 0.0_um;
-  do {
+  for (int i = 1;; ++i) {
     std::transform(pos_moving.begin(), pos_moving.end(), do_step(i).begin(), pos_moving.begin(), std::plus<float>());
     distance_edge = pitch - std::abs(pos_moving[displ_ind]);
-    // current diffusion value
-    double sigma = std::sqrt(i * diffusion_step / distance0) * (distance0 / thickness) * sigma0;
+    // Current diffusion value
+    const double sigma = 0.4_um;
```
Comment (civanch) [inline SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.cc:173]: @claralasa , inside the loop there is a mixture of double and float number, operations, and constants. Is it possible use only float, for example, use 0.5f. Can constants be defined before the loop.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by srimanob** [comment]
@claralasa The short matrix does not include geometry with 3D pixel. Is T25 (D80) or T26 (D81) OK for the test? I will retrigger the test again.

By the way, as warning shown in
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d62f29/14232/llvm-analysis/esrget-sa.txt
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d62f29/14232/llvm-analysis/cmsclangtidy.txt
Do you have plan for migration?

FYI @mmusich

**Comment by mmusich** [comment]
@srimanob 

>The short matrix does not include geometry with 3D pixel. Is T25 (D80) or T26 (D81) OK for the test?

Yes, these are OK, but T23 (D79, wf: 35834.0) is even better (it has 3D also in the endcaps).

> Do you have plan for migration?

Well,  for the `getByToken` warnings in clang-tidy as I understand it's only "suggestion" for the moment, and will be processed in due time.  For the `esConsumes` these are treated "centrally" already by the core group in PR https://github.com/cms-sw/cmssw/pull/31697. What's the plan about that one?

**Comment by srimanob** [comment]
Please test workflow 35834.0

**Comment by srimanob** [comment]
> > The short matrix does not include geometry with 3D pixel. Is T25 (D80) or T26 (D81) OK for the test?
> 
> Yes, these are OK, but T23 (D79, wf: 35834.0) is even better (it has 3D also in the endcaps).
Thanks, I trigger the test with 35834.0.

> > Do you have plan for migration?
> 
> Well, for the `getByToken` warnings in clang-tidy as I understand it's only "suggestion" for the moment, and will be processed in due time. For the `esConsumes` these are treated "centrally" already by the core group in PR #31697. What's the plan about that one?

OK, thanks. Then I will ignore the warning for now, and will follow up with the core team.

**Comment by srimanob** [comment]
+Upgrade

**Comment by qliphy** [comment]
+1

**Pull request merged**